### DISCUSSION TITLE
CISC와 RISC

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#261] [Computer Architecture]
.

**Comment by 3Juhwan** [comment]
| 특징                    | CISC (Complex Instruction Set Computer) | RISC (Reduced Instruction Set Computer)          |
|-------------------------|-----------------------------------------|--------------------------------------------------|
| **명령어 집합**         | 복잡하고 다양한 명령어                  | 단순하고 제한된 명령어                           |
| **명령어 길이**         | 가변 길이 명령어 사용                   | 고정 길이 명령어 사용                            |
| **명령어 실행 시간**    | 명령어에 따라 실행 시간이 다름           | 대부분의 명령어가 1 클럭 내외로 실행됨          |
| **파이프라인 적합성**   | 파이프라인 구현이 어려움                | 파이프라이닝에 적합                              |
| **메모리 접근 방식**    | 다양한 메모리 접근 방식                | 로드(load)와 스토어(store) 명령어만 사용        |
| **레지스터 활용**       | 상대적으로 레지스터 수가 적음            | 많은 범용 레지스터를 적극적으로 활용            |
| **주소 지정 방식**      | 다양한 주소 지정 방식 사용               | 간단하고 제한된 주소 지정 방식                  |
| **프로그램 코드 길이**  | 적은 명령어로 복잡한 작업 수행          | 더 많은 명령어 필요                              |
| **메모리 절약**         | 메모리 공간을 절약할 수 있음            | 프로그램 코드가 더 길어져 메모리 사용 증가      |
| **하드웨어 설계 복잡성**| 복잡한 하드웨어 설계                    | 단순한 하드웨어 설계                             |
| **역사적 배경**         | 메모리 제약이 크던 과거에 많이 사용     | 파이프라인 성능을 중시하는 현대 설계에 적합     |

**Comment by le2sky** [comment]
명령어의 모양, 명령어로 할 수 있는 연산, 주소 지정 방식은 CPU 마다 차이가 있다. 이때 각 CPU가 이해할 수 있는 명령어들의 모음을 **명령어 집합(instruction set architecture)** 이라고 한다. 

 ex) 같은 프로그램이어도 ISA 에따라 어셈블리어도 달라진다.

명령어 병렬 처리 기법에는 명령어 파이프라인, 슈퍼스칼라, 비순차적 명령어 처리 방식이 있다. 이러한 명령어 병렬 처리 기법을 적용하기 용이한 ISA가 있는 반면, 그렇지 않은 ISA가 존재한다. 대표적인 ISA인 CISC와 RISC를 기준으로 알아보자.

**CISC(Complex Instruction Set Computer)**

CISC는 확장 명령형 컴퓨터 혹은 복합 명령 집합 컴퓨터라는 의미로 복합적인 명령어를 통해 연산한다는 의미이다.

- CISC는 다양하고 강력한 기능의 명령어 집합을 활용한다. 따라서, 명령어의 형태와 크기가 다양한 **가변 길이 명령어**를 활용한다. 메모리에 접근하는 주소 지정 방식도 다양하다. 
- 다양하고 강력한 명령어를 활용하기 때문에 상대적으로 적은 수의 명령어도로 프로그램을 실행할 수 있다. 따라서, 메모리 공각을 절약할 수 있다는 장점이 있다.
- (단점) 명령어의 크기와 실행 시간이 일정하지 않기 때문에 명령어 파이프라인에 적합하지 않다. 
- (단점) 대다수의 복잡한 명령어는 사용 빈도가 낮다. 

**RISC(Reduced Instruction Set Computer)**
RISC는 축소 명령형 컴퓨터라는 의미로 명령어의 종류를 적게 하고, 내부회로를 단순하게 만들어 명령어 자체를 고속으로 실행할 수 있도록 향상 시킨 아키텍쳐이다. RISC는 CISC에 비해 상대적으로 명령어의 종류가 적으며, 각 명령어는 짧고 규격화되어 있으며 되도록 1클럭 내외로 실행되는 명령어를 지향한다.

- **고정 길이 명령어**를 활용한다.
- 명령어가 규격화되어 있고, 하나의 명령어가 1클럭 내외로 실행되기 때문에 명령어 파이프라이닝 최적화에 유리하다.
- 메모리 접근 방식을 load, store 두 개로 제한한다. (대신 레지스터를 적극적으로 활용)
- CISC보다 레지스터 이용 연산이 많고, 일반적인 경우 범용 레지스터 개수도 많다 
- (단점?) CISC보다 더 많은 명령어로 동일한 프로그램을 실행한다.
